### PR TITLE
Make access to the java consumer fair

### LINKED ISSUE #1107: PartitionnedStream hanging after stream consumption started
Hi, I notice a bug in `v2.6.0` that is not in `v2.3.4`.

I am using a `Consumer.partitionedStream` and for each partition I need to get the committed and endOffset with the `Consumer.endOffsets` / `Consumer.committed` method: 

```
Consumer.partitionedStream(???).mapZIOParUnordered(12)(
        (e: (TopicPartition, stream.Stream[Throwable, CommittableRecord[Bytes, Either[UnknownEvent, Event]]])) => {
          val (topicPartition, streamIn) = e
          Consumer.endOffsets(Set(topicPartition)).flatMap {
            ZIO.debug("some more exec here") *> streamIn.runDrain
          }
        }
      )
```
I have 12 partitions so I use `mapZIOParUnordered` with a parallelism of 12 to process all my partitions.

I notice that as soon as the first stream starts for one partition, then the Consumer.endOffsets and Consumer.committed are really long for others partitions and too long. It looks like the polling of the consumer blocks other threads to achieve Consumer.endOffsets and Consumer.committed tasks. The result after few minutes is that all the partitions are not assigned anymore forever but the process is not kill.

### PR DESCRIPTION
**Pull request by flavienbert**
Replace the semaphore that is used to control access to the java consumer with a reentrant lock that has fairness enabled. This prevents liveness problems for consumers that need to use the consumer, e.g. to fetch committed offsets. Fixes #1107.

### CODE REVIEW

**Comment by flavienbert** [comment]
Fixes #1107

**Comment by svroonland** [comment]
Have you verified that this fixes the issue? A unit test would of course be very welcome

File: zio-kafka/src/main/scala/zio/kafka/consumer/internal/ConsumerAccess.scala [hunk 1]
```diff
@@ -8,16 +8,18 @@ import zio.kafka.consumer.ConsumerSettings
...
-  access: Semaphore
+  access: ReentrantLock
 ) {
+
   def withConsumer[A](f: ByteArrayKafkaConsumer => A): Task[A] =
     withConsumerZIO[Any, A](c => ZIO.attempt(f(c)))
 
   def withConsumerZIO[R, A](f: ByteArrayKafkaConsumer => RIO[R, A]): RIO[R, A] =
-    access.withPermit(withConsumerNoPermit(f))
+    access.lock *> withConsumerNoPermit(f) <* access.unlock
```
Comment (erikvanoosten) [inline zio-kafka/src/main/scala/zio/kafka/consumer/internal/ConsumerAccess.scala:22]: Shouldn't that be:
```suggestion
    access.lock.zipRight.withConsumerNoPermit(f).ensuring(access.unlock)
```

**Comment by erikvanoosten** [comment]
I like it!

So now the question is, how to unit test this? Perhaps by calling to the consumer from the stream (e.g. get offsets), and then verifying that it makes progress?

**Comment by erikvanoosten** [comment]
> Probably the java consumer is not used in time during shutdow process?

Likely.

File: zio-kafka/src/main/scala/zio/kafka/consumer/internal/ConsumerAccess.scala (hunk 1 above)
Comment (erikvanoosten) [reply zio-kafka/src/main/scala/zio/kafka/consumer/internal/ConsumerAccess.scala:22]: Or perhaps this is better:
```suggestion
    (access.lock *> withConsumerNoPermit(f)).ensuring(access.unlock)
```

**Comment by flavienbert** [comment]
I tried to run the test I wrote with the semaphore version, and the test fail.

**Review by erikvanoosten — APPROVED** [review]

**Comment by erikvanoosten** [comment]
Thanks @flavienbert !

**Comment by flavienbert** [comment]
@erikvanoosten you're welcome. Do you plan to publish a new version with this fix soon?

**Comment by erikvanoosten** [comment]
Yes definitely! Probably the coming weekend.

File: build.sbt [hunk 2]
```diff
@@ -100,7 +100,8 @@ lazy val zioKafka =
     .settings(
       libraryDependencies ++= Seq(
         kafkaClients,
-        scalaCollectionCompat
+        scalaCollectionCompat,
+        "dev.zio" %% "zio-concurrent" % zioVersion.value
```
Comment (guizmaii) [inline build.sbt:104]: Why do we need to add this dependency? zio-concurrent's repo has been archived: https://github.com/zio/zio-concurrent. I'm not sure it's a good idea to be dependent on it. 🤔

**Comment by erikvanoosten** [comment]
> @flavienbert @erikvanoosten Seeing this PR from ZIO (merged but not yet released): [zio/zio#8554](https://github.com/zio/zio/pull/8554) and seeing that zio-concurrent repo is archived, do we really want/need to use zio-concurrent `ReentrantLock`?
> 
> Edit: I checked and `ReentrantLock` is directly present in ZIO. We don't need the zio-concurrent dependency

Zio-concurrent is now a module in the main zio repository.
Eh, I do not see it in ZIO itself... did I miss something?

**Comment by guizmaii** [comment]
@erikvanoosten My bad. I though it was coming from the zio-concurrent repo but comes indeed from ZIO's main repo

**Comment by erikvanoosten** [comment]
Now that the standard semaphore in zio is fair (since 2.0.20, see https://github.com/zio/zio/pull/8554), we can revert this change, allowing us to remove an external dependency.

**Pull request merged**